day3/src/main.rs

```diff
@@ -0,0 +1,59 @@
+use std::{collections::HashSet, fs::read_to_string};
+
+fn main() {
+    let input = read_file("input.txt");
+    println!("{:?}", puzzle1(&input));
+    println!("{:?}", puzzle2(&input));
+}
+
+fn puzzle1(input: &str) -> u32 {
+    input
+        .split("\n")
+        .map(|rucksack| {
+            let (first_half, second_half) = rucksack.split_at(rucksack.len() / 2);
+            let first_half_letters_set: HashSet<char> = HashSet::from_iter(first_half.chars());
+
+            // find the common letter between first_half and second_half
+            for char in second_half.chars() {
+                if first_half_letters_set.contains(&char) {
+                    return priority(char);
+                }
+            }
+            0
```

Comment:
If we want to keep this strictly functional, it's quite weird to place a `for` loop (which is assumed to have side effects) in a `map`. We can instead use the built in `find` that returns an `Option`, and use the `map_or` to unwrap that option and act on it:

```suggestion
            second_half
                .chars()
                .find(|c| first_half_letters_set.contains(c))
                .map_or(0, priority)
```